File: Week1/validParentheses/validParentheses.js
//algorithm in python solution


const isValid = s => {
    pairs = {
        "]": "[",
        "}": "{",
        ")": "("
    }

    let openStack = [];

    for (let i = 0; i < s.length ; i++) {
        const currentParenthesis = s[i];
        
        //check to see if it is open
        if (pairs[currentParenthesis] === undefined) {
            openStack.push(currentParenthesis)
        } else {
            //check to see if closed isn't at beginning
            if (openStack.length !== 0) {
                //check to if corresponding open is the last item in openStack
                if (pairs[currentParenthesis] === openStack[openStack.length - 1]) {
                    openStack.pop();
                } else {
                    return false;
                }
            } else {
                return false;
            }
        }
    }

    
    if (openStack.length !== 0) {
        return false
    } else {
        return true;
    }
}

console.log(isValid("(])"))

File: Week1/validParentheses/valid_parentheses.py
#create an pairs dictionary, key closed parenthesis, value open
#create a openStack
#traverse through the input string. 
#check if it is closed or open. 
#if open:
    #add to openStack
#if close:
    #check if length of stack is not 0 (to make sure closed is not at the beginning of string)
        #check if the corresponding open bracket is the last item is stack:
            # pop last item from stack
        #else: (not the last item in stack corresponds to he closed bracket)
            # return false
    # else: (closed is at the beginning of sentence)
    #     return false



def is_valid_parentheses(s):
    pairs = {
        ')': '(',
        ']': '[',
        '}': '{'
    }

    open_stack = []

    for p in s:
        #if it is open
        if p not in pairs:
            open_stack.append(p)
        #if it is a closed bracket
        else:
            if len(open_stack) != 0:
                if pairs[p] == open_stack[-1]:
                    open_stack.pop()
                else:
                    return False
            else:
                return False
    
    if len(open_stack) == 0:
        return True 
    else:
        return False      


print(is_valid_parentheses("(])"))

